### PR TITLE
RYA-104 Mongo Rya Shell Integration

### PR DESCRIPTION
**Pull request by kchilton2**
## Description
Added Mongo support to the Rya Shell.

### Tests
Added tests.

### Links
https://issues.apache.org/jira/browse/RYA-104

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/585/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/586/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/587/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector' /><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/587/org.apache.rya$rya.prospector/testReport'>incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/587/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.mr/ProspectorTest/testCount/'><strong>org.apache.rya.prospector.mr.ProspectorTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/587/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/587/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testNoAuthsCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testNoAuthsCount</strong></a></li></ul>

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/596/

File: common/rya.api/src/main/java/org/apache/rya/api/client/RyaClient.java [hunk 1]
```diff
@@ -53,37 +55,37 @@
...
+        this.deletePeriodicPcj = requireNonNull(deletePeriodicPcj);
+        this.listIncrementalQueries = requireNonNull(listIncrementalQueries);
         this.bactchUpdatePCJ = requireNonNull(batchUpdatePcj);
         this.getInstanceDetails = requireNonNull(getInstanceDetails);
         this.instanceExists = requireNonNull(instanceExists);
         this.listInstances = requireNonNull(listInstances);
         this.addUser = requireNonNull(addUser);
         this.removeUser = requireNonNull(removeUser);
         this.uninstall = requireNonNull(uninstall);
-        this.loadStatements = requireNonNull(loadStatements);
```
Comment (meiercaleb) [inline common/rya.api/src/main/java/org/apache/rya/api/client/RyaClient.java]: Why did the null checks get removed?

File: extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if(username.isPresent()) {
+            conf.setMongoUser(username.get());
+        }
+
+        if(password.isPresent()) {
+            conf.setMongoPassword( new String(password.get()) );
+        }
+
+        // Default to having a table prefix of "rya", otherwise the app doesn't work for some reason.
+        conf.setTablePrefix("rya");
```
Comment (meiercaleb) [inline extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java]: That is strange behavior.  You don't have any idea where it breaks?  Was the table prefix hard coded somewhere else?  Seems like hard coding it here just perpetuates the problem...

File: extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoUninstall.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     * @param instanceExists - The interactor used to check if a Rya instance exists. (not null)
+     */
+    public MongoUninstall(final MongoClient adminClient, final MongoInstanceExists instanceExists) {
+        this.adminClient = requireNonNull(adminClient);
+        this.instanceExists = requireNonNull(instanceExists);
+    }
+
+    @Override
+    public void uninstall(final String ryaInstanceName) throws InstanceDoesNotExistException, RyaClientException {
+        try {
```
Comment (meiercaleb) [inline extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoUninstall.java]: Preconditions

File: extras/rya.pcj.fluo/pom.xml [hunk 4]
```diff
@@ -37,7 +37,7 @@
         <module>pcj.fluo.api</module>
         <module>pcj.fluo.app</module>
         <module>pcj.fluo.client</module>
-        <module>pcj.fluo.integration</module>
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pom.xml]: This probably shouldn't be commented out.

File: extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java (hunk 2 above)
Comment (kchilton2) [reply extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java]: This requirement is hard coded in RyaSailFactory for unknown reasons, so I'm working around it. It is probably a defect.

File: common/rya.api/src/main/java/org/apache/rya/api/client/RyaClient.java (hunk 1 above)
Comment (kchilton2) [reply common/rya.api/src/main/java/org/apache/rya/api/client/RyaClient.java]: They shouldn't have been. Adding them back in.

File: extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoUninstall.java (hunk 3 above)
Comment (kchilton2) [reply extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoUninstall.java]: Done.

File: extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java (hunk 2 above)
Comment (kchilton2) [reply extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoConnectionDetails.java]: It looks like there's already a ticket open for this. 
https://issues.apache.org/jira/browse/RYA-135

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/607/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/606/

**Pull request merged**